### PR TITLE
add xpu option to enable advanced path

### PR DESCRIPTION
**Pull request by Dewei-Wang-sh**

### CODE REVIEW

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp [hunk 1]
```diff
@@ -68,6 +68,10 @@ struct ConvertTritonGPUToLLVM
     : public triton::gpu::intel::impl::ConvertTritonIntelGPUToLLVMBase<
           ConvertTritonGPUToLLVM> {
   using ConvertTritonIntelGPUToLLVMBase::ConvertTritonIntelGPUToLLVMBase;
+  ConvertTritonGPUToLLVM() = default;
+  ConvertTritonGPUToLLVM(bool advancedPath) {
+    this->advancedPath = advancedPath;
+  }
```
Comment (victor-eds) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp:74]:
```suggestion
  ConvertTritonGPUToLLVM(bool advancedPath) 
      : advancedPath(advancedPath) {}
```

File: third_party/intel/backend/compiler.py [hunk 2]
```diff
@@ -56,6 +56,7 @@ class XPUOptions:
     backend_name: str = 'intel'
     sanitize_overflow: bool = False
     generate_native_code: bool = False
+    advanced_path: bool = False
```
[the remark this answers is not in the document]
Comment (Dewei-Wang-sh) [reply third_party/intel/backend/compiler.py:59]: it's not good, 
1. the original ones is not alpha order
2. this is newly added for xpu, good to put it at the back

File: benchmarks/triton_kernels_benchmark/flash_attention_fwd_benchmark.py [hunk 3]
```diff
@@ -214,6 +214,7 @@ def forward(q, k, v, causal, sm_scale):
             num_warps=num_warps,  #
             num_stages=num_stages,  #
             grf_mode='large',  #
+            advanced_path=True,  #
```
[the remark this answers is not in the document]
Comment (Dewei-Wang-sh) [reply benchmarks/triton_kernels_benchmark/flash_attention_fwd_benchmark.py:217]: not enabled, still guarded by an env var, just a backup mechanism.

File: third_party/intel/backend/compiler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Dewei-Wang-sh) [reply third_party/intel/backend/compiler.py:59]: it's not good, 
1. the original ones is not alpha order
2. this is newly added for xpu, good to put it at the back

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp (hunk 1 above)
Comment (Dewei-Wang-sh) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp:74]: won't work.

File: third_party/intel/backend/compiler.py [hunk 4]
```diff
@@ -232,8 +233,9 @@ def make_ttgir(mod, metadata, opt, properties):
         pm = ir.pass_manager(mod.context)
         pm.enable_debug()
 
-        if (properties["has_subgroup_2d_block_io"] and properties["has_subgroup_matrix_multiply_accumulate"]
-                and os.getenv("TRITON_INTEL_ADVANCED_PATH", "0") == "1"):
+        if (os.getenv("TRITON_INTEL_ADVANCED_PATH", "0") == "1" or opt.advanced_path):
+            if not (properties["has_subgroup_2d_block_io"] and properties["has_subgroup_matrix_multiply_accumulate"]):
```
Comment (whitneywhtsang) [inline third_party/intel/backend/compiler.py:237]: Do we need to worry that with this change, we need different source code for different platforms?

Comment (Dewei-Wang-sh) [reply third_party/intel/backend/compiler.py:237]: https://github.com/intel/intel-xpu-backend-for-triton/pull/2732#discussion_r1850539401
maybe we should keep the original way.

**Review by victor-eds — APPROVED** [review]

File: third_party/intel/backend/compiler.py (hunk 4 above)
Comment (whitneywhtsang) [reply third_party/intel/backend/compiler.py:237]: How about only assert that in PipelineManager.h and keep the original way in compiler.py and TritonGPUToLLVM.cpp?

Comment (Dewei-Wang-sh) [reply third_party/intel/backend/compiler.py:237]: good idea, let me assert in the gputollvm and pass that to pipelinemanager.

**Pull request merged**